### PR TITLE
bugfix/12165-12163-button-width-changed-fontweight

### LINKED ISSUE #12165: Renderer.button with auto width and fixed height
#### Expected behaviour
Rendered button without errors in console.

#### Actual behaviour
Using `Renderer.button()` with auto `width` and fixed `height` throws errors.

#### Live demo with steps to reproduce
- Renderer: http://jsfiddle.net/BlackLabel/z4p15r3v/
- RangeSelector + style: http://jsfiddle.net/BlackLabel/w8z9xjop/

### PR DESCRIPTION
**Pull request by sebastianbochan**
Fixed #12165, #12163, width of button was wrong when changed fontweight.

### CODE REVIEW

**Comment by sebastianbochan** [comment]
The differences in tests are caused by situation from ticket: https://github.com/highcharts/highcharts/issues/12163 where fontWeight did not have influence on width of button.

**Comment by TorsteinHonsi** [comment]
> The differences in tests are caused by situation from ticket: #12163 where fontWeight did not have influence on width of button and proper align.

That must mean that `getBBox` was called twice for these labels and that it was cached the first time 🤔 Can you do a little digging and see why it was called twice? The `getBBox` function is generally slow and should be used with care. This may actually have to do with one of my comments above, that `updateBoxSize` is called unnecessarily.

File: ts/parts/SvgRenderer.ts [hunk 1]
```diff
@@ -6343,6 +6347,13 @@ extend(SVGRenderer.prototype, /** @lends Highcharts.SVGRenderer.prototype */ {
                         updateBoxSize();
                         updateTextPadding();
                     }
+
+                    // Keep updated (#12163)
```
[the remark this answers is not in the document]
Comment (sebastianbochan) [reply ts/parts/SvgRenderer.ts]: Indeed, I was sure that '100px' is valid, but as I see is not.

File: ts/parts/SvgRenderer.ts [hunk 2]
```diff
@@ -6066,9 +6067,13 @@ extend(SVGRenderer.prototype, /** @lends Highcharts.SVGRenderer.prototype */ {
             bBox = ((
                 typeof width === 'undefined' ||
                 typeof height === 'undefined' ||
+                // #12165 error when width is null (auto)
+                width === null ||
+                height === null ||
```
[the remark this answers is not in the document]
Comment (sebastianbochan) [reply ts/parts/SvgRenderer.ts]: Indeed, I was sure that '100px' is valid, but is not.

File: ts/parts/SvgRenderer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (sebastianbochan) [reply ts/parts/SvgRenderer.ts]: Optymized that and reduced the number of `updateBoxSize` calling.

**Comment by sebastianbochan** [comment]
It turned out that I called it twice (on width and on fontWeight) which was wrong. Now I optymized the process and also skip any calculations / check the condition, when font style does not exist.

File: ts/parts/SvgRenderer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (TorsteinHonsi) [reply ts/parts/SvgRenderer.ts]: Looking good now, thanks!

**Comment by sebastianbochan** [comment]
Ok, will do that. At the beginning I will focus on Highstock (because is simple reason of problem - in the past, bolded text did not have influence on width). 

I will let you know if hesitate about the any test.

**Comment by sebastianbochan** [comment]
@TorsteinHonsi 

9 tests remains.
   - 7 are correct in my opinion, but always is better to ensure
   - 2 have no "source" to compare. So I cannot observe the difference. I guess that its the same difference like in others with rangeselector.

**Pull request merged**